drivers: add imx7d i2c driver

diegosueiro opened the pull request:
This pull request is to add i2c driver for mcimx7_m4 arch and colibri_imx7d_m4 board.

The implementation was tested on the colibri_imx7d_m4 board by compiling and running the samples/grove/lcd application (extra delays between messages have to be added because of the grove lcd module restrictions) using the I2C_4 bus. Another validation was made with a simple application performing reads and writes to an RTC (m41t0) located in the Toradex Colibri Evaluation board.

galak commented on drivers/i2c/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+zephyr_sources_if_kconfig(i2c_imx.c)
```
Seems like we should keep this the same as the other lines:

zephyr_source_ifdef(CONFIG_I2C_IMX . i2c_imx.c)

galak commented on drivers/i2c/Kconfig.imx:
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig IMX_I2C
```
Don't add a new Kconfig file, please model this off of what is done for i2c_mcux, use CONFIG_I2C_1, CONFIG_I2C_2, etc.

diegosueiro replied:
For imx we have the hardware mapped from I2C_1 to I2C_4 and in the Kconfig we have I2C_0 to I2C_3.

What is your suggestion, add the I2C_4 option to Kconfig or use 0 to 3 and consider 0 as 1, 1 as 2 and so on, for the imx?

galak replied:
I'm saying add I2C_4 to Kconfig, and just use I2C_1..I2C_4

galak commented on drivers/i2c/i2c_imx.c:
```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright (c) 2018 NXP
```
Shouldn't this be your copyright?

galak reviewed (dismissed):
See comments, also need to add device tree support.

galak commented on dts/bindings/i2c/nxp,imx-i2c.yaml:
```diff
@@ -0,0 +1,33 @@
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "nxp,imx-i2c"
```
should be compatible = "fsl,imx7d-i2c";

diegosueiro replied:
@galak,

This binding can be used by other imx cores (ie. imx6sx_m4). I believe that it is the same situation of nxp,imx-gpio.yaml, that shares the same dts binding and gpio driver shim.

I even think that they can share the same i2c driver shim. Maybe @stanislav-poboril can help us in checking this.

diegosueiro replied:
So, do you think we have to change in nxp,imx-i2c.yaml to ```constraint: "fsl,imx7d-i2c", "fsl,imx21-i2c"``` and on nxp_imx7d_m4.dtsi ```compatible = "fsl,imx7d-i2c"```?

Sorry but I got a little bit confused what needs to be changed.

diegosueiro replied:
If I change from "nxp,imx-i2c" to "fsl,imx21-i2c" I'll have to change NXP_IMX_I2C to FSL_IMX21_I2C in the dts.fixup file which will not follow the definition pattern of the other devices like UART and GPIO and looks strange to me.

galak replied:
> So, do you think we have to change in nxp,imx-i2c.yaml to constraint: "fsl,imx7d-i2c", "fsl,imx21-i2c" and on nxp_imx7d_m4.dtsi compatible = "fsl,imx7d-i2c"?
> 
> Sorry but I got a little bit confused what needs to be changed.

So we don't handle multiple compats well today with regards to the binding files.  So I'm saying we use 'fsl,imx7d-i2c' so that we are matching linux (at least partially).

I understand that the pattern will differ from other items, but that is ok.  This is what should be in the DTS for the i2c controller.

MaureenHelm replied:
> So we don't handle multiple compats well today with regards to the binding files.

Does this mean we can define exactly one constraint in a given yaml file, but we can define multiple compatibles in the dts?

galak replied:
> Does this mean we can define exactly one constraint in a given yaml file, but we can define multiple compatibles in the dts?

Correct, the dts is happy to have multiple compats.  The yaml for the binding don't handle this well.  Right now there is an expectation of 1:1 yaml to compat.

diegosueiro replied:
@galak ,

Devicetree compatible string suggestion addressed.

dbkinder reviewed (approved):
+1 for doc changes

diegosueiro committed:
arch: Add imx7d_m4 i2c definitions

Adds all necessary i2c definitions and configurations for imx7d_m4 soc.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@galak and @MaureenHelm ,

Are you expecting to merge this before the merge window close?

diegosueiro commented:
@galak and @MaureenHelm,

Should I rebase this PR to get it merged?

diegosueiro commented on drivers/i2c/i2c_imx.c:
```diff
@@ -0,0 +1,454 @@
+	do {
+
+		/* Forget old flags except start flag */
+		flags &= I2C_MSG_RESTART;
```
@MaureenHelm ,

I've followed the same approach implemented in i2c_bitbang.c and tested with different i2c messages combinations and it worked without any issues.Preserving the restart flag is used to decide if it needs to send and start (by setting the appropriate mode according to the data sheet) or repeat start on the bus.

diegosueiro replied:
Done.

diegosueiro commented on boards/arm/colibri_imx7d_m4/Kconfig.board:
```diff
@@ -9,3 +9,5 @@ config BOARD_COLIBRI_IMX7D_M4
 	bool "Toradex Colibri iMX7 Dual"
 	depends on SOC_SERIES_IMX7_M4
 	select SOC_PART_NUMBER_MCIMX7D5EVM10SC
+	select HAS_DTS_I2C if I2C
```
@galak,

What should I do with the HAS_DTS_I2C selection?

galak replied:
Seems like the driver should select this, not the board.

diegosueiro replied:
@MaureenHelm and @galak ,

Done.

diegosueiro commented on boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts:
```diff
@@ -18,6 +18,7 @@
 		uart-2 = &uart2;
 		led0   = &green_led;
 		sw0    = &user_switch_1;
+		i2c_4 = &i2c4;
```
Done

diegosueiro commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -109,6 +111,10 @@ was tested with the following pinmux controller configuration.
 +---------------+-----------------+---------------------------+
+| SODIMM_194    | I2C4_SDA        | I2C_SDA                   |
++---------------+-----------------+---------------------------+
+| SODIMM_196    | I2C3_SCL        | I2C_SCL                   |
```
Done

diegosueiro commented on drivers/i2c/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 zephyr_library()
 
+zephyr_library_sources_ifdef(CONFIG_I2C_IMX		i2c_imx.c)
```
Done

diegosueiro committed:
board: add i2c support for colibri_imx7d_m4 board

Adds definitions, configurations and device tree entries for
colibri_imx7d_m4 board.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

MaureenHelm commented on drivers/i2c/i2c_imx.c:
```diff
@@ -0,0 +1,451 @@
+	do {
+
+		/* Start condition? */
+		if (flags & I2C_MSG_RESTART) {
```
This condition will only be satisfied on the first iteration and doesn't need to be inside the do/while loop. If you move it out, then you can move `flags |= msgs->flags` to the beginning of the loop, and use `flags` consistently throughout the loop. The current mix of `flags` and `msgs->flags` is confusing.

diegosueiro replied:
Done

MaureenHelm commented on drivers/i2c/Kconfig:
```diff
@@ -202,6 +202,14 @@ config I2C_MCUX
+	bool "i.MX I2C driver"
+	depends on HAS_IMX_I2C
+	select HAS_DTS_I2C
+	default n
```
Remove this line. See #8586

diegosueiro replied:
Done

MaureenHelm commented on ext/hal/nxp/imx/Kconfig:
```diff
@@ -28,4 +28,10 @@ config HAS_IMX_GPIO
 
+config HAS_IMX_I2C
+	bool
+	default n
```
Remove this line

diegosueiro replied:
Done

diegosueiro committed:
drivers: add i.MX I2C driver shim

Adds a shim layer around the imx i2c driver to adapt it to the Zephyr
i2c master interface.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
